src/org/traccar/events/OverspeedEventHandler.java

```diff
@@ -50,8 +55,48 @@ public OverspeedEventHandler() {
 
         Collection<Event> events = new ArrayList<>();
         double speed = position.getSpeed();
+        double speedLimit = 0;
 
-        if (globalSpeedLimit != 0 && speed > globalSpeedLimit) {
+        if (device.getAttributes().containsKey(ATTRIBUTE_SPEED_LIMIT)) {
+            speedLimit = Double.parseDouble((String) device.getAttributes().get(ATTRIBUTE_SPEED_LIMIT));
+        }
+        if (speedLimit == 0 && checkGroups) {
+            long groupId = device.getGroupId();
+            while (groupId != 0) {
+                if (Context.getDeviceManager().getGroupById(groupId).getAttributes()
+                        .containsKey(ATTRIBUTE_SPEED_LIMIT)) {
+                    speedLimit = Double.parseDouble((String) Context.getDeviceManager().getGroupById(groupId)
+                            .getAttributes().get(ATTRIBUTE_SPEED_LIMIT));
+                    if (speedLimit != 0) {
+                        break;
+                    }
+                }
+                if (Context.getDeviceManager().getGroupById(groupId) != null) {
+                    groupId = Context.getDeviceManager().getGroupById(groupId).getGroupId();
+                } else {
+                    groupId = 0;
+                }
+            }
+        }
+        if (speedLimit == 0) {
+            try {
+                Server server = Context.getDataManager().getServer();
+                if (server.getAttributes().containsKey(ATTRIBUTE_SPEED_LIMIT)) {
+                    speedLimit = Double.parseDouble((String) server.getAttributes().get(ATTRIBUTE_SPEED_LIMIT));
+                }
+            } catch (SQLException error) {
+                Log.warning(error);
+            }
+        }
+        double oldSpeed = 0;
+        if (riseOnce) {
+            Position lastPosition = Context.getDeviceManager().getLastPosition(position.getDeviceId());
+            if (lastPosition != null) {
+                oldSpeed = lastPosition.getSpeed();
+            }
+        }
+        speedLimit = UnitsConverter.knotsFromKph(speedLimit);
+        if (speedLimit != 0 && speed > speedLimit && oldSpeed <= speedLimit) {
```

Comment:
Thanks. It applies for all events. I guess we can have another event filtering handler (which can cache old events or just type and timestamp or something).